### DISCUSSION TITLE
Facebook openid driver

### DISCUSSION THREAD

**Discussion by gkielwasser** [directus/directus#9744] [Archive]
Hi,

I'm trying to set Facebook authentication using the openid driver. 
I have set the following envs.
```
AUTH_FACEBOOK_DRIVER=openid
AUTH_FACEBOOK_SCOPE="openid"
AUTH_FACEBOOK_CLIENT_ID="string:$CLIENT_ID"
AUTH_FACEBOOK_CLIENT_SECRET="$CLIENT_SECRET"
AUTH_FACEBOOK_ISSUER_URL="https://www.facebook.com/.well-known/openid-configuration"
```

Also I have seen [this ticket](https://github.com/directus/directus/pull/9598) that seemed to fix authentication with openid on Facebook.

I'm testing it on Directus v9.0.1

Here is the exception that is returned by `openid-cient` library
```
token_endpoint must be configured on the issuer
```

Does anyone can help about that?

Thanks

**Comment by aidenfoxx** [comment]
For some reason Facebook only supports the `id_token` flow with OpenID and we use the more secure `code` flow. Someone could adapt our implementation to fallback to the `id_token` flow if `code` isn't supported... but that won't be me.

You can use the OAuth2 driver with the following URLs for now, and it should work:

```
AUTH_FACEBOOK_AUTHORIZE_URL="https://www.facebook.com/dialog/oauth"
AUTH_FACEBOOK_ACCESS_URL="https://graph.facebook.com/oauth/access_token"
AUTH_FACEBOOK_PROFILE_URL="https://graph.facebook.com/me?fields=email"
```

**Reply by gkielwasser** [reply]
> For some reason Facebook only supports the `id_token` flow with OpenID and we use the more secure `code` flow. Someone could adapt our implementation fallback to the `id_token` flow if `code` isn't supported... but that won't be me.
> 
> You can use the OAuth2 driver with the following URLs for now, and it should work:
> 
> ```
> AUTH_FACEBOOK_AUTHORIZE_URL="https://www.facebook.com/dialog/oauth"
> AUTH_FACEBOOK_ACCESS_URL="https://graph.facebook.com/oauth/access_token"
> AUTH_FACEBOOK_PROFILE_URL="https://graph.facebook.com/me"
> ```

Thanks, I have made the change that should be appropriated:
```
AUTH_FACEBOOK_DRIVER=oauth2
AUTH_FACEBOOK_SCOPE="email"
AUTH_FACEBOOK_CLIENT_ID="string:$CLIENT_ID"
AUTH_FACEBOOK_CLIENT_SECRET="$CLIENT_SECRET"
AUTH_FACEBOOK_AUTHORIZE_URL="https://www.facebook.com/dialog/oauth"
AUTH_FACEBOOK_ACCESS_URL="https://graph.facebook.com/oauth/access_token"
AUTH_FACEBOOK_PROFILE_URL="https://graph.facebook.com/me"
```

I have now the following error
```
Failed to find user identifier for provider "facebook"
Invalid user credentials.
```

**Reply by gkielwasser** [reply]
I have managed to have it working with
```
AUTH_FACEBOOK_IDENTIFIER_KEY="id" 
```
instead of 
```
AUTH_FACEBOOK_IDENTIFIER_KEY="email" 
```

With the email identifier it happen that the userInfo retrieved doesn't contains the "email" even if the asked scope is "email"

**Comment by gkielwasser** [comment]
My bad ! Asking for the scope email is not enough.
The PROFILE_URL has to be update to ask the email field.
```
AUTH_FACEBOOK_PROFILE_URL="https://graph.facebook.com/me?fields=email
```

**Reply by aidenfoxx** [reply]
Oh nice! Good solution. I'll update my comment to reflect that for anyone else trying to figure this out! 😄

**Reply by benhaynes** [reply]
Anything we can add to the docs to clarify or warn? People will look there before here. @aidenfoxx

**Reply by aidenfoxx** [reply]
It could be a warning in the docs. But the same could be said about any OAuth quirk that we've still yet to find, since every OAuth implementation can differ.

We do have warnings in the logs when providers aren't working as expected, and 99% of services should work with some degree of configuration tweaking. But we can't document everything, so it's tricky. Maybe just document the more common quirks.

**Comment by benhaynes** [comment]
Gotcha. More common quirks, or more common providers, makes sense to document. 👍

**Comment by eikaramba** [comment]
for anyone stumbling accross this discussion. this config worked for me, after having one issue after another

AUTH_FACEBOOK_DRIVER="oauth2"
AUTH_FACEBOOK_SCOPE="openid"
AUTH_FACEBOOK_IDENTIFIER_KEY="id" 
AUTH_FACEBOOK_CLIENT_ID="string:XXXXXXXXXXXXXX"
AUTH_FACEBOOK_CLIENT_SECRET="XXXXXXXXXXXXXXXX"
AUTH_FACEBOOK_AUTHORIZE_URL="https://www.facebook.com/dialog/oauth"
AUTH_FACEBOOK_ACCESS_URL="https://graph.facebook.com/oauth/access_token"
AUTH_FACEBOOK_PROFILE_URL="https://graph.facebook.com/me?fields=email"
AUTH_FACEBOOK_ALLOW_PUBLIC_REGISTRATION="true"
AUTH_FACEBOOK_DEFAULT_ROLE_ID="XXXXXXXXXXX" #only if you need this
AUTH_FACEBOOK_ICON="facebook"
